python3Packages.limits: fix input tarball non-reproducibility

delroth opened the pull request:
###### Description of changes

Fixes regression from last bump (22f5d327101db), it looks like the
release tarball was changed upstream.

Tested: `nix-build . -A home-assistant` now passes on staging-next.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

delroth commented:
cc @fabaff

dotlambda commented:
> it looks like the release tarball was changed upstream.

Can we please tell upstream that's not okay?

fabaff commented:
It could be that I messed things up with the hash.

delroth commented:
> It could be that I messed things up with the hash.

I'm still investigating, but ofborg was fine with that hash according to the logs, and it doesn't match any other version I can find still published.

> Can we please tell upstream that's not okay?

I'll file an issue when I've confirmed it's on their side. I suggest not blocking on this so staging-next can get greener faster? :)

delroth commented:
Successful fetch on ofborg:

```
trying https://github.com/alisaifee/limits/archive/2.6.1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   133  100   133    0     0    413      0 --:--:-- --:--:-- --:--:--   413
100 88449    0 88449    0     0  87918      0 --:--:--  0:00:01 --:--:-- 87918
```

Failed fetch on Hydra:

```
trying https://github.com/alisaifee/limits/archive/2.6.1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 88440    0 88440    0     0   150k      0 --:--:-- --:--:-- --:--:--  150k
```

The archive size seems to have changed.

dotlambda commented:
See https://github.com/NixOS/nixpkgs/issues/84312

dotlambda commented:
This is how I solved it for another package but I don't think it will work in this case: https://github.com/NixOS/nixpkgs/blob/d21c94cd35ab90d237a768cbce4a1376c339f705/pkgs/development/python-modules/pikepdf/default.nix#L41

delroth committed:
pythonPackages.limits: hack around versioneer non-reproducibility

Context: #84312 documents that using .gitattributes on GitHub results in
non-reproducible release tarballs. versioneer uses this to provide
version numbers for Python libraries.

Hack around that issue by deleting the file containing substitutions in
the fixed hash input derivation, and recreate the required info from the
derivation's version info at patchPhase time.

delroth commented:
@dotlambda thanks for the reference, and there are actually a few other Python packages using versioneer that have similar workarounds. PTAL.

delroth commented:
@ofborg build python39Packages.limits python310Packages.limits
